### ISSUE TITLE
Removing None from variable type of arguments to functions 

### ISSUE THREAD

**Issue by danielgordon10** [microsoft/pyright#2310]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

```python
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.axes import Axes
def render(goal: np.ndarray, ax: Optional[Axes] = None) -> Axes:
    if ax is None:
        fig = plt.figure()
        ax = fig.gca()
    ax.plot(goal[0, 0], goal[0, 1], "X", color="g", markersize=15)  # results in "plot" is not a known member of "None"
    return ax
```
However
```python
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.axes import Axes
def render(goal: np.ndarray, ax: Optional[Axes] = None) -> Axes:
    ax_ = ax
    if ax_ is None:
        fig = plt.figure()
        ax_ = fig.gca()
    ax_.plot(goal[0, 0], goal[0, 1], "X", color="g", markersize=15)  # no error
    return ax_
```
This is one such occurrence but I'm sure you could make an even simpler one.

**Comment by erictraut** [comment]
Could you provide a complete stand-alone example that defines all of the symbols? I presume that `np` is an import of `numpy`, but I don't know where `plt` or `Axes` come from.

**Comment by danielgordon10** [comment]
Sorry about that, fixed.

**Comment by erictraut** [comment]
Where are you getting type info for the `matplotlib` library? Are you using the [matplotlib type stubs](https://github.com/microsoft/python-type-stubs/tree/main/matplotlib) that ship with pylance? Or are you relying on `useLibraryCodeForTypes` to infer types from the library code? I'm guessing it's the latter.

The difference you're seeing here is not because `ax` is a parameter and `_ax` is a local variable. It's due to the fact that `ax` is annotated and `_ax` is not.  If you annotate `_ax` with the type `Optional[Axes]`, you will see the same behavior.

When pyright performs assignment-based type narrowing for an annotated type, it doesn't narrow to `Any` or `Unknown` even if the assigned value is `Any` or `Unknown`. It assumes that the caller is returning a type that is compatible with the annotated type and does not attempt to narrow it. Because `fig` has an `Unknown` type, the expression `fig.gca()` is also `Unknown`, and assigning `Unknown` to `ax` does not narrow the type. Its type after the `if` statement is therefore `Axes | None`.

In the case of an unannotated variable, the type of the variable is the union of all types assigned to it. By design, pyright is more permissive in this case. In your example, `ax_` is unannotated. It is assigned values in two locations. In the first location, it is assigned a value of type `Axes | None`, which is narrowed to just `Axes` in the implied `else` branch of the `if` statement. In the second case, it is assigned a type of `Unknown`. After the `if` statement, its type is therefore `Axes | Unknown`.

So, this is expected behavior.

You can avoid this by using type stubs for matplotlib.

**Comment by danielgordon10** [comment]
Gotcha. Thanks for looking into it.
